sklearn/linear_model/ridge.py

```diff
@@ -1004,9 +1129,18 @@ def _errors_and_values_helper(self, alpha, y, v, Q, QT_y):
         We don't construct matrix G, instead compute action on y & diagonal.
         """
         w = 1. / (v + alpha)
-        constant_column = np.var(Q, 0) < 1.e-12
-        # detect constant columns
-        w[constant_column] = 0  # cancel the regularization for the intercept
+        if self.fit_intercept:
+            # the vector containing the square roots of the sample weights (1
+            # when no sample weights) is the eigenvector of XX^T which
+            # corresponds to the intercept; we cancel the regularization on
+            # this dimension. the corresponding eigenvalue is
+            # sum(sample_weight).
+            if self._with_sw:
+                intercept_dim = np.isclose(
+                    Q, self._normalized_sqrt_sw[:, None]).all(axis=0)
+            else:
+                intercept_dim = np.var(Q, 0) < 1.e-12
```

Comment:
I changed it to compare to what the singular vector we are looking for is, ie +- normalized vector of ones, to be consistent to the case with sample weights.
it is indeed possible to look for the vector with the smallest angle rather than testing with `np.isclose`, do you think it is better?

Reply:
I changed it to cancel regularization on the singular vector which has the smallest angle (modulo pi) with the direction of the intercept